pydantic/_internal/_fields.py

```diff
@@ -277,3 +213,45 @@ def _is_finalvar_with_default_val(type_: type[Any], val: Any) -> bool:
         return False
     else:
         return True
+
+
+def collect_dataclass_fields(
+    cls: type[StandardDataclass], types_namespace: dict[str, Any] | None, *, typevars_map: dict[Any, Any] | None = None
+) -> dict[str, FieldInfo]:
+    from ..fields import FieldInfo
+
+    fields: dict[str, FieldInfo] = {}
+    type_hints = get_cls_type_hints_lenient(cls, types_namespace)
+    dataclass_fields: dict[str, dataclasses.Field] = cls.__dataclass_fields__
+
+    for ann_name, dataclass_field in dataclass_fields.items():
+        ann_type = type_hints[ann_name]
+        if is_classvar(ann_type):
+            continue
+
+        if not dataclass_field.init:
+            # TODO: We should probably do something with this so that validate_assignment behaves properly
+            #   See https://github.com/pydantic/pydantic/issues/5470
```

Comment:
I think this will be fairly involved so I don't want to fix it in this PR. (It's an issue on main, not just this branch.)